source/features/show-whitespaces.tsx

```diff
@@ -0,0 +1,96 @@
+import './show-whitespaces.css';
+import React from 'dom-chef';
+import select from 'select-dom';
+import features from '../libs/features';
+import onPrFileLoad from '../libs/on-pr-file-load';
+import onNewComments from '../libs/on-new-comments';
+import getTextNodes from '../libs/get-text-nodes';
+
+function showWhiteSpacesOn(line: Element): void {
+	const textNodes = getTextNodes(line);
+
+	for (const textNode of textNodes) {
+		const nodeValue = textNode.textContent!;
+		if (nodeValue.length === 0) {
+			continue;
+		}
+
+		const fragment = document.createDocumentFragment();
+
+		let lastEncounteredCharType;
+		let charType: 'space' | 'tab' | 'other';
+		let node;
+
+		for (const char of nodeValue) {
+			if (char === ' ') {
+				charType = 'space';
+			} else if (char === '\t') {
+				charType = 'tab';
+			} else {
+				charType = 'other';
+			}
+
+			if ((lastEncounteredCharType && lastEncounteredCharType === charType) && node) {
+				node.textContent += char;
+
+				if (charType === 'space') {
+					node.dataset.rghSpaces += '·';
+				} else if (charType === 'tab') {
+					node.dataset.rghTabs += '→';
+				}
+			} else {
+				if (node) {
+					fragment.append(node);
+				}
+
+				if (charType === 'space') {
+					node = <span className="pl-ws pl-space" data-rgh-spaces="·">{char}</span>;
+				} else if (charType === 'tab') {
+					node = <span className="pl-ws pl-tab" data-rgh-tabs="→">{char}</span>;
+				} else {
+					node = <>{char}</>;
+				}
```

Comment:
I think that if you put this here, you can drop it from both places (line 42 and line 58):

```js
if (node) {
	fragment.append(node);
}
```

Reply:
`<>{char}</>` creates a `#document-fragment` that cannot be modified once appended to the DOM. The append operation has to happen only on the next iteration when the current fragment is filled with required characters (hence the additional check after the end of loop).

Reply:
In that case, I think you can just use `else { fragment.append(char) }`, right?

Reply:
Using `.append()` creates a new `#text` node for each appended character, that was the reason I had to shift to using `.textContent += ` in https://github.com/sindresorhus/refined-github/commit/5d7c3d31d8c593ad98ace45f4047a70af52e6a9b#diff-0f8cf70b012a85b8fddc8744630b8e96R34.